### PR TITLE
Modify `git elegant feature` command

### LINKED ISSUE #32: Move changes from current to new branch
Problem:
- make some changes
- call `git elegant feature xxx`
- error: `Your index contains uncommitted changes.`

Solution:
- create a stash (`git stash`) if there are some changed
-  remove the stash after moving to the new branch (`git stash pop`)

Add unit tests to verify positive and negative cases.

### PR DESCRIPTION
**Pull request by ivan-podorozhnyi**
Updates make this command work properly in case of merge conflicts.
Added one negative unit test.
Closes #32

### CODE REVIEW

**Comment by ivan-podorozhnyi** [comment]
@extsoft Please make it clear about command exit code call in:
```
_validate() {
    if [ -z "$1" ]; then
        __rm "$2 is not set"
        exit -1
    fi
}
```
From http://www.tldp.org/LDP/abs/html/exitcodes.html I found that exit code -1 can't be checked because exit takes only integer args in the range 0 - 255. What was your reason to use `exit -1`?

File: src/main/git-elegant [hunk 1]
```diff
@@ -102,8 +102,10 @@ push-after-rebase() {
 
 feature() {
     _validate "$1" "Feature name"
+    git stash
```
Comment (AntonKarabaza) [inline src/main/git-elegant]: @ivan-podorozhnyi Seems that here and on `git stash pop` we need some checks, because command `feature` might be used with no changes on branch, so `git stash` won't create stash and `git stash pop` will apply some unsuitable stash.

Comment (ivan-podorozhnyi) [reply src/main/git-elegant]: @AntonKarabaza Ok, I will add some checks.

**Comment by extsoft** [comment]
> What was your reason to use exit -1?

@ivan-podorozhnyi no reasons. Let's define some correct exit code here.

**Comment by ivan-podorozhnyi** [comment]
@extsoft Let's define rules for exit codes in #73

**Comment by ivan-podorozhnyi** [comment]
@AntonKarabaza Added some checks, take a look.

File: src/main/git-elegant-feature [hunk 2]
```diff
@@ -3,6 +3,14 @@ set -e
...
 default() {
     _validate "$1" "Feature name"
+    local status="git status | grep 'Changes to be commited: || Changes not staged for commit:'"
+    if [ -n $(eval "$status") ]; then
+        git stash
+    fi
+
     git-elegant pull $MASTER
     git checkout -b "$1"
+    if [[ $(git stash list) ]]; then
```
Comment (AntonKarabaza) [inline src/main/git-elegant-feature]: @ivan-podorozhnyi On this check might be some mess, if no stash was added but list contain some another stashes.

**Comment by ivan-podorozhnyi** [comment]
@AntonKarabaza I added a flag, looks like now its much better.

File: src/test/git-elegant-feature.bats [hunk 3]
```diff
@@ -8,9 +8,25 @@ setup() {
     fake-pass git "fetch --tags"
     fake-pass git pull
     fake-pass git "checkout -b test-feature"
+    fake-pass git "stash"
+    fake-pass git "stash pop"
+    fake-pass git "status | grep 'Changes to be commited: || Changes not
+    staged for commit:'" "not empty"
+    fake-pass git "stash list" "message"
```
Comment (AntonKarabaza) [inline src/test/git-elegant-feature.bats]: @ivan-podorozhnyi This mock is redundant.

**Comment by ivan-podorozhnyi** [comment]
@AntonKarabaza Done, removed it.

**Comment by AntonKarabaza** [comment]
Good job!

File: src/main/git-elegant-feature [hunk 4]
```diff
@@ -2,7 +2,17 @@
 set -e
 
 default() {
+    local uncommited_changes=false
     _validate "$1" "Feature name"
+    local status="git status | grep 'Changes to be commited: || Changes not staged for commit:'"
+    if [ -n $(eval "$status") ]; then
+        git stash
```
Comment (extsoft) [inline src/main/git-elegant-feature]: We need to set some name for a stash:
```
git stash save "git-elegant: Save changes to unstash them to '$1' branch"
```

File: src/test/git-elegant-feature.bats [hunk 5]
```diff
@@ -8,9 +8,24 @@ setup() {
     fake-pass git "fetch --tags"
     fake-pass git pull
     fake-pass git "checkout -b test-feature"
+    fake-pass git "stash"
+    fake-pass git "stash pop"
+    fake-pass git "status | grep 'Changes to be commited: || Changes not
```
Comment (extsoft) [inline src/test/git-elegant-feature.bats]: Incorrect mock. You have to mock each command between `|` separately.

File: src/test/git-elegant-feature.bats [hunk 6]
```diff
@@ -8,9 +8,24 @@ setup() {
...
+    staged for commit:'" "not empty"
 }
 
-@test "exit code is 0 when run 'git-elegant feature'" {
+@test "exit code is 0 when run 'git-elegant feature test-feature'" {
   run git-elegant feature test-feature
   [ "$status" -eq 0 ]
 }
+
```
Comment (extsoft) [inline src/test/git-elegant-feature.bats]: Please add tests for all possible executions (have changes and not).

File: src/main/git-elegant-feature (hunk 4 above)
Comment (ivan-podorozhnyi) [reply src/main/git-elegant-feature]: @extsoft Do you want to use this stash name?
>git stash save "git-elegant: Save changes to unstash them to '$1' branch"

Maybe something shorter? For example:
```
git stash save "stash made by git-elegant"
```

**Comment by ivan-podorozhnyi** [comment]
@extsoft Please take a look at my recent updates. Tests were added.

File: src/main/git-elegant-feature [hunk 7]
```diff
@@ -2,7 +2,17 @@
 set -e
 
 default() {
+    local uncommited_changes=false
     _validate "$1" "Feature name"
+    local status="git status | grep 'Changes to be commited: || Changes not staged for commit:'"
```
Comment (extsoft) [inline src/main/git-elegant-feature]: There is a bug: git elegant will stash even there are no changes. Please create unit test, then fix.

Log:
```bash
~/data/elegant-git git status
On branch stash3
Untracked files:
  (use "git add <file>..." to include in what will be committed)

	src/main/test

nothing added to commit but untracked files present (use "git add" to track)
~/data/elegant-git git status | grep 'Changes to be commited: || Changes not staged for commit:'
~/data/elegant-git status="git status | grep 'Changes to be commited: || Changes not staged for commit:'"
~/data/elegant-git if [ -n $(eval "$status") ]; then echo "ff"; fi
ff
```

File: src/test/git-elegant-feature.bats [hunk 8]
```diff
@@ -8,9 +8,37 @@ setup() {
...
+@test "print message when run 'git-elegant feature'" {
+  run git-elegant feature
+  [[ "${lines[0]}" =~ "Feature name is not set" ]]
+}
+
+@test "exit code is 0 when run 'git-elegant feature' with changes" {
+  fake-pass git status "Changes to be commited:"
+  run git-elegant feature test-feature
+  [ "$status" -eq 0 ]
+
```
Comment (extsoft) [inline src/test/git-elegant-feature.bats]: Spare line.

File: src/test/git-elegant-feature.bats [hunk 9]
```diff
@@ -8,9 +8,37 @@ setup() {
...
+}
+
+@test "exit code is 0 when run 'git-elegant feature' without changes" {
+  fake-pass git status
+  run git-elegant feature test-feature
+  [ "$status" -eq 0 ]
+
```
Comment (extsoft) [inline src/test/git-elegant-feature.bats]: Spare line.

**Comment by ivan-podorozhnyi** [comment]
@extsoft Please take a look at updated command implementation.

File: src/main/git-elegant-feature [hunk 10]
```diff
@@ -1,8 +1,26 @@
 #!/usr/bin/env bash
 set -e
 
+_check_changes(){
```
Comment (extsoft) [inline src/main/git-elegant-feature]: From `man grep`:
```
EXIT STATUS
     The grep utility exits with one of the following values:

     0     One or more lines were selected.
     1     No lines were selected.
     >1    An error occurred.
```

Please remove this function.

**Comment by ivan-podorozhnyi** [comment]
@extsoft Changes applied:
- no `git status` called anymore, now `git stash` is used to check the changes;
- one if/else used in algorithm; 
- added all possible unit tests.

File: src/test/git-elegant-feature.bats [hunk 11]
```diff
@@ -8,9 +8,34 @@ setup() {
...
+  [ "$status" -eq 255 ]
+}
+
+@test "print message when run 'git-elegant feature'" {
+  run git-elegant feature
+  [[ "${lines[0]}" =~ "Feature name is not set" ]]
+}
+
+@test "exit code is 0 when run 'git-elegant feature' with changes" {
+  fake-pass git status "Saved working directory"
```
Comment (extsoft) [inline src/test/git-elegant-feature.bats]: @ivan-podorozhnyi I think you fake-pass the wrong command, is it? The same about line#38.

Comment (extsoft) [reply src/test/git-elegant-feature.bats]: Also, may a good idea is to `fake-fail` `git stash drop ..` with a specific exit code and check that this code appears in the end. it will guarantee that unstash logic was invoked.

Comment (ivan-podorozhnyi) [reply src/test/git-elegant-feature.bats]: > I think you fake-pass the wrong command, is it? The same about line#38.

@extsoft Sorry, my bad. Resolved.

> Also, may a good idea is to fake-fail git stash drop .. with a specific exit code and check that this code appears in the end. it will guarantee that unstash logic was invoked.

@extsoft I added a unit test to check it. Please, take a look.

But another thing that I found is that in case of merge conflicts our app closes without dropping stashed changes. Do you think it's normal behavior? Should we apply some changes or add a separate issue to discuss it?

Comment (extsoft) [reply src/test/git-elegant-feature.bats]: > But another thing that I found is that in case of merge conflicts our app closes without dropping stashed changes.

@ivan-podorozhnyi the idea is to store changes for a customer. If we can apply them, we haven't delete.

**Comment by extsoft** [comment]
@rultor please merge
